### PR TITLE
KAFKA-6987 Reimplement KafkaFuture with CompletableFuture

### PR DESCRIPTION
**Pull request by andrasbeni**
Use CompletableFuture in KafkaFutureImpl to implement KafkaFuture.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
cc @cmccabe

**Comment by ijuma** [comment]
Can `KafkaFuture` implement `CompletionStage`?

**Comment by andrasbeni** [comment]
@ijuma I think `KafkaFuture` can implement `CompletionStage`. 

Don't mind below comment since I figured out we don't need any additional execution facility as we already have a CompleteableFuture that we can use.

--------------------------------------------------------------------------------------------
There are CompletionStage.*Async(...) methods that do not accept an executor but, as their documentation says, 

> Returns a new CompletionStage that, when this stage completes  normally, is executed using this stage's default asynchronous  **execution facility**, with this stage's result as the argument to  the supplied function.

(emphasis added)

So this means we need to supply that "execution facility" maybe in the form of 
 * using ForkJoinPool.commonPool()
 * or firing up a new background thread for each task (which is what CompletableFuture does in some circumstances) 

A full blown ThreadPoolExecutor in clients, I think, would be an overkill 
If you agree with one of the above options, I am happy to jump on implementing it.

**Comment by cmccabe** [comment]
Thanks for looking at this, @andrasbeni.  I think the challenge here is the handling of exceptions in thenApply and whenComplete.

The documentation says:
> if a stage's computation terminates abruptly with an (unchecked) exception or error, then all dependent stages requiring its completion complete exceptionally as well, with a CompletionException holding the exception as its cause.

However, KafkaFuture was never able to implement this, because CompletionException is a jdk8 only class.  I think in order to make this work, we need to add some glue code which unwraps the CompletionException to preserve the existing behavior.  Otherwise everyone handling exceptions will break, unfortunately.

**Comment by cmccabe** [comment]
By the way, I don't know if you mentioned it anywhere, but one thing we could do is have an accessor which exposed the CompletableFuture object inside KafkaFuture.  That would allow people to use any CF methods they wanted directly, without us adding shim functions.  Or integrate well with APIs that expect CF specifically

**Comment by andrasbeni** [comment]
@cmccabe Thanks for raising the issue with whenComplete and thenApply. I tried to capture our custom behavior in a few tests to make sure I implement the same logic. 

@ijuma The problem with implementing `CompletionStage` is `whenComplete`. Right now user can write
`new KafkaFutureImpl<String>().thenApply(String::length)`. The lambda expression will be compiled to a `KafkaFuture.BiConsumer`. If we implement `CompletionStage`, there will be another `whenComplete` in the same class that takes `java.util.function.BiConsumer`, another  SAM interface, as its parameter. That's why compiler will report ambiguity and will fail to compile above code.

**Comment by andrasbeni** [comment]
@ijuma , @cmccabe can you please take a look?

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 1]
```diff
@@ -16,250 +16,135 @@
...
-        return true;
+        return completableFuture.completeExceptionally(newException);
     }
 
     /**
      * If not already completed, completes this future with a CancellationException.  Dependent
      * futures that have not already completed will also complete exceptionally, with a
      * CompletionException caused by this CancellationException.
+     *
+     * TODO was this true?
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:135]: This TODO should be removed

File: clients/src/main/java/org/apache/kafka/common/KafkaFuture.java [hunk 2]
```diff
@@ -202,4 +207,13 @@ public abstract T get(long timeout, TimeUnit unit) throws InterruptedException,
...
     @Override
     public abstract boolean isDone();
+
+    /**
+     * Returns a ComletableFuture equivalent to this Future.
+     *
+     * Implemented in {@link KafkaFuture} throws UnsuportedOperationException.
+     */
+    public CompletableFuture<T> toCompletableFuture() {
+        throw new UnsupportedOperationException();
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/common/KafkaFuture.java:217]: Why isn't this function abstract?

**Comment by cmccabe** [comment]
@andrasbeni, I took a first pass over this and it looks good.  Probably the best reason to do this is that it would let us use more stuff from CompletableFuture over time.

I am a bit concerned about the error handling, though.  Does CompletableFuture really wrap cancellation exception in completion exception?  I had trouble figuring this out from the Java documentation.  I think in general we need a lot more tests of the old and new exception handling.  For example, complete something exceptionally through the underlying CF and test what effects that has on `KafkaFuture#whenComplete`, etc.

`copyWith` should never have been added -- it was added due to a misunderstanding of how `whenComplete` worked, as I recall.  I thought we removed it before the release, but I guess not.  This might be a case where we should actually make use of that "Evolving" annotation and just remove it.

**Pull request closed without merging**